File: 백준/레벨Silver I/11660. 구간 합 구하기 5/구간 합 구하기 5.cc
#include <bits/stdc++.h>

using namespace std;

void rangeSum() {
    // 누적 합 초기화
    int N, M;
    cin >> N >> M;

    vector<vector<int>> nums(N, vector<int>(N, 0));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> nums[i][j];
        }
    }    

    vector<vector<int>> pre(N+1, vector<int>(N+1, 0));
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            pre[i][j] = pre[i-1][j] + pre[i][j-1] - pre[i-1][j-1] + nums[i-1][j-1];
        }
    }      

    // 구간 합 처리
    while (M-- > 0) {
        int x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;

        int t = pre[x2][y2] - pre[x1-1][y2] - pre[x2][y1-1] + pre[x1-1][y1-1];
        cout << t << "\n";
    }      
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    rangeSum();
}

File: 백준/레벨Silver I/11660. 구간 합 구하기 5/구간 합 구하기 5.py
from sys import stdin

input = stdin.readline

def range_sum():
    # 누적 합 초기화
    N, M = map(int, input().split())

    nums = []
    for _ in range(N):
        nums.append(list(map(int, input().split())))

    pre = [[0] * (N + 1) for _ in range(N + 1)]
    for i in range(1, N+1):
        for j in range(1, N+1):
            pre[i][j] = pre[i-1][j] + pre[i][j-1] - pre[i-1][j-1] + nums[i-1][j-1]
    
    # 구간 합 처리
    for _ in range(M):
        x1, y1, x2, y2 = map(int, input().split())

        t = pre[x2][y2] - pre[x1-1][y2] - pre[x2][y1-1] + pre[x1-1][y1-1]
        print(t)

range_sum()